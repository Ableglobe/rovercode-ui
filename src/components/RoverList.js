import React, { Component, Fragment } from 'react';
import {
  Button,
  Card,
  Confirm,
  Form,
  Grid,
  Header,
  Icon,
  Loader,
  Modal,
  Segment,
} from 'semantic-ui-react';
import { Link, Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';

import CustomPagination from './CustomPagination';

const defaultState = {
  confirmOpen: false,
  newRoverOpen: false,
  newRoverName: null,
  newRoverId: null,
  focusRover: {
    id: null,
    name: null,
  },
};

class RoverList extends Component {
  constructor(props) {
    super(props);

    this.state = defaultState;
  }

  componentDidMount() {
    const { fetchRovers } = this.props;

    return fetchRovers();
  }

  showConfirm = e => this.setState({
    confirmOpen: true,
    focusRover: {
      id: e.target.id,
      name: e.target.name,
    },
  })

  cancelRemove = () => this.setState(defaultState)

  removeRover = () => {
    const { fetchRovers, removeRover } = this.props;
    const { focusRover } = this.state;

    this.setState(defaultState);

    return removeRover(focusRover.id).then(() => fetchRovers());
  }

  modalButton = () => (
    <Button primary style={{ marginLeft: '10px' }} onClick={this.handleNewRoverOpen}>
      <Icon name="plus" />
      Register New Rover
    </Button>
  )

  handleNewRoverOpen = () => this.setState({ newRoverOpen: true })

  handleNewRoverClose = () => this.setState({ newRoverOpen: false })

  handleNameChange = e => this.setState({ newRoverName: e.target.value })

  createRover = () => {
    const { createRover } = this.props;
    const { newRoverName } = this.state;

    this.setState(defaultState);

    return createRover({ name: newRoverName }).then(newRover => this.setState({
      newRoverId: newRover.value.id,
    }));
  }

  handlePageChange = (e, { activePage }) => {
    const { fetchRovers } = this.props;

    return fetchRovers(activePage);
  }

  render() {
    const { isFetching, rovers } = this.props;
    const {
      confirmOpen,
      focusRover,
      newRoverOpen,
      newRoverId,
    } = this.state;

    return (
      <Fragment>
        {
          newRoverId ? (
            <Redirect to={{
              pathname: `/rovers/${newRoverId}`,
              state: { created: true },
            }}
            />
          ) : (null)
        }
        <Modal trigger={this.modalButton()} open={newRoverOpen} onClose={this.handleNewRoverClose}>
          <Modal.Header>
            Enter the name of the rover
          </Modal.Header>
          <Modal.Content>
            <Form id="nameForm" onSubmit={this.createRover}>
              <Form.Input placeholder="Rover name" onChange={this.handleNameChange} required />
            </Form>
          </Modal.Content>
          <Modal.Actions>
            <Button primary type="submit" form="nameForm">
              Create
            </Button>
            <Button onClick={this.handleNewRoverClose}>
              Cancel
            </Button>
          </Modal.Actions>
        </Modal>
        {
          isFetching || rovers === null
            ? (<Loader active />)
            : (
              <Segment raised style={{ margin: '10px' }}>
                <Header as="h1" textAlign="center">
                  Rovers
                </Header>
                <Card.Group centered>
                  {
                    rovers.results.map(rover => (
                      <Card key={rover.id}>
                        <Card.Content>
                          <Card.Header>
                            {rover.name}
                          </Card.Header>
                        </Card.Content>
                        <Card.Content extra>
                          <Button primary as={Link} to={`/rovers/${rover.id}`}>
                              Configure
                          </Button>
                          <Button
                            negative
                            id={rover.id}
                            name={rover.name}
                            onClick={this.showConfirm}
                            floated="right"
                          >
                              Remove
                          </Button>
                        </Card.Content>
                      </Card>
                    ))
                  }
                </Card.Group>
                <Grid centered>
                  <Grid.Row>
                    <CustomPagination
                      defaultActivePage={1}
                      totalPages={rovers.total_pages}
                      onPageChange={this.handlePageChange}
                    />
                  </Grid.Row>
                </Grid>
                <Confirm
                  header="Remove Rover"
                  content={`Are you sure you want to remove ${focusRover.name}?`}
                  open={confirmOpen}
                  onConfirm={this.removeRover}
                  onCancel={this.cancelRemove}
                  cancelButton="No"
                  confirmButton="Yes"
                />
              </Segment>
            )
        }
      </Fragment>
    );
  }
}

RoverList.defaultProps = {
  rovers: {
    next: null,
    previous: null,
    total_pages: 1,
    results: [],
  },
  isFetching: false,
};

RoverList.propTypes = {
  fetchRovers: PropTypes.func.isRequired,
  removeRover: PropTypes.func.isRequired,
  createRover: PropTypes.func.isRequired,
  rovers: PropTypes.shape({
    next: PropTypes.string,
    previous: PropTypes.string,
    total_pages: PropTypes.number,
    results: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
      }),
    ),
  }),
  isFetching: PropTypes.bool,
};

export default RoverList;
