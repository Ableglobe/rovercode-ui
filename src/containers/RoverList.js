import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { createRover, fetchRovers, removeRover } from '../actions/rover';
import { updateValidAuth } from '../actions/auth';
import RoverList from '../components/RoverList';

const mapStateToProps = ({ rover }) => ({ ...rover });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  fetchRovers: () => {
    const fetchRoversAction = fetchRovers({
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(fetchRoversAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
  createRover: (settings) => {
    const createRoverAction = createRover(settings, {
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(createRoverAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
  removeRover: (id) => {
    const removeRoverAction = removeRover(id, {
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(removeRoverAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
});

const RoverListContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RoverList);

RoverListContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(RoverListContainer));
