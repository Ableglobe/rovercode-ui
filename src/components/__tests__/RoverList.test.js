import React from 'react';
import { MemoryRouter, Redirect } from 'react-router';
import {
  Button,
  Card,
  Form,
  Header,
  Loader,
} from 'semantic-ui-react';
import { mountWithIntl, shallowWithIntl } from 'enzyme-react-intl';
import CustomPagination from '../CustomPagination';
import RoverList from '../RoverList';

let createRover;
let fetchRovers;
let removeRover;

describe('The RoverList component', () => {
  beforeEach(() => {
    createRover = jest.fn(() => Promise.resolve({
      value: {
        id: 1,
      },
    }));
    fetchRovers = jest.fn(() => Promise.resolve({}));
    removeRover = jest.fn(() => Promise.resolve({}));
  });

  test('renders on the page with no errors', () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches rovers on mount', async () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    fetchRovers = jest.fn();
    await mountWithIntl(
      <MemoryRouter>
        <RoverList
          rovers={rovers}
          createRover={createRover}
          fetchRovers={fetchRovers}
          removeRover={removeRover}
        />
      </MemoryRouter>,
    );
    expect(fetchRovers.mock.calls.length).toBe(1);
  });

  test('shows the correct number of rovers for the user', async () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();
    await wrapper.instance().componentDidMount();
    wrapper.update();

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Loader).exists()).toBe(false);

    expect(wrapper.find(Card).length).toBe(2);
    expect(wrapper.find(Card).first().find(Button).first()
      .prop('to')).toBe('/rovers/1');
    expect(wrapper.find(Card.Header).first().prop('children')).toBe('Sparky');

    expect(wrapper.find(Card).last().find(Button).first()
      .prop('to')).toBe('/rovers/2');
    expect(wrapper.find(Card.Header).last().prop('children')).toBe('Marvin');
  });

  test('shows no rovers on error', async () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();
    await wrapper.instance().componentDidMount();
    wrapper.update();

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Loader).exists()).toBe(false);
  });

  test('shows loader when fetching', () => {
    const wrapper = shallowWithIntl(
      <RoverList
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
        isFetching
      />,
    ).dive();

    expect(wrapper.find(Header).exists()).toBe(false);
    expect(wrapper.find(Loader).exists()).toBe(true);
  });

  test('creates a rover and goes to new rover detail', async () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.setState({
      newRoverName: 'Rovey',
    });
    await wrapper.instance().createRover();
    wrapper.update();

    expect(wrapper.find(Redirect).exists()).toBe(true);
    expect(wrapper.find(Redirect).prop('to')).toEqual({
      pathname: '/rovers/1',
      state: {
        created: true,
      },
    });
    expect(fetchRovers).toHaveBeenCalledTimes(1);
    expect(createRover).toHaveBeenCalledWith({
      name: 'Rovey',
    });
  });

  test('removes a rover and reloads the rover list', async () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.setState({
      focusRover: {
        id: 1,
        name: 'Sparky',
      },
    });
    await wrapper.instance().removeRover();

    expect(fetchRovers).toHaveBeenCalledTimes(2);
    expect(removeRover).toHaveBeenCalledWith(1);
  });

  test('shows confirm dialog', () => {
    const wrapper = shallowWithIntl(
      <RoverList
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.instance().showConfirm({
      target: {
        parentNode: {
          id: undefined,
          name: undefined,
        },
        id: 1,
        name: 'Sparky',
      },
    });

    expect(wrapper.state('confirmOpen')).toBe(true);
  });

  test('cancel dialog does not remove rover', () => {
    const wrapper = shallowWithIntl(
      <RoverList
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.instance().cancelRemove();

    expect(fetchRovers).toHaveBeenCalledTimes(1);
    expect(removeRover).not.toHaveBeenCalled();
    expect(wrapper.state('confirmOpen')).toBe(false);
  });

  test('shows new rover dialog', () => {
    const wrapper = shallowWithIntl(
      <RoverList
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.instance().handleNewRoverOpen();

    expect(wrapper.state('newRoverOpen')).toBe(true);

    wrapper.find(Form.Input).simulate('change', {
      target: {
        value: 'Sparky',
      },
    });

    expect(wrapper.state('newRoverName')).toBe('Sparky');
  });

  test('cancel dialog does not create rover', () => {
    const wrapper = shallowWithIntl(
      <RoverList
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.instance().handleNewRoverClose();

    expect(fetchRovers).toHaveBeenCalledTimes(1);
    expect(removeRover).not.toHaveBeenCalled();
    expect(wrapper.state('newRoverOpen')).toBe(false);
  });

  test('change page on click', () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 2,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    wrapper.find(CustomPagination).simulate('pageChange', null, {
      activePage: 2,
    });

    expect(fetchRovers).toHaveBeenCalledWith(2);
  });

  test('no pagination for single page', () => {
    const rovers = {
      next: null,
      previous: null,
      total_pages: 1,
      results: [{
        id: 1,
        name: 'Sparky',
        client_id: '1234',
      }, {
        id: 2,
        name: 'Marvin',
        client_id: '5678',
      }],
    };
    const wrapper = shallowWithIntl(
      <RoverList
        rovers={rovers}
        createRover={createRover}
        fetchRovers={fetchRovers}
        removeRover={removeRover}
      />,
    ).dive();

    expect(wrapper.find(CustomPagination).exists()).toBe(false);
  });
});
