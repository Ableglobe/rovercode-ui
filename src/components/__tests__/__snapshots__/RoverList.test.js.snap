// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RoverList component renders on the page with no errors 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RoverList
    createRover={[MockFunction]}
    fetchRovers={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": Promise {},
          },
        ],
      }
    }
    removeRover={[MockFunction]}
    rovers={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        null,
        <Modal
          centered={true}
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          onClose={[Function]}
          open={false}
          trigger={
            <Button
              as="button"
              onClick={[Function]}
              primary={true}
              style={
                Object {
                  "marginLeft": "10px",
                }
              }
            >
              <Icon
                as="i"
                name="plus"
              />
              Register New Rover
            </Button>
          }
        >
          <ModalHeader>
            Enter the name of the rover
          </ModalHeader>
          <ModalContent>
            <Form
              as="form"
              id="nameForm"
              onSubmit={[Function]}
            >
              <FormInput
                as={[Function]}
                control={[Function]}
                onChange={[Function]}
                placeholder="Rover name"
                required={true}
              />
            </Form>
          </ModalContent>
          <ModalActions>
            <Button
              as="button"
              form="nameForm"
              primary={true}
              type="submit"
            >
              Create
            </Button>
            <Button
              as="button"
              onClick={[Function]}
            >
              Cancel
            </Button>
          </ModalActions>
        </Modal>,
        <Loader
          active={true}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "centered": true,
          "children": Array [
            <ModalHeader>
              Enter the name of the rover
            </ModalHeader>,
            <ModalContent>
              <Form
                as="form"
                id="nameForm"
                onSubmit={[Function]}
              >
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Rover name"
                  required={true}
                />
              </Form>
            </ModalContent>,
            <ModalActions>
              <Button
                as="button"
                form="nameForm"
                primary={true}
                type="submit"
              >
                Create
              </Button>
              <Button
                as="button"
                onClick={[Function]}
              >
                Cancel
              </Button>
            </ModalActions>,
          ],
          "closeOnDimmerClick": true,
          "closeOnDocumentClick": false,
          "dimmer": true,
          "eventPool": "Modal",
          "onClose": [Function],
          "open": false,
          "trigger": <Button
            as="button"
            onClick={[Function]}
            primary={true}
            style={
              Object {
                "marginLeft": "10px",
              }
            }
          >
            <Icon
              as="i"
              name="plus"
            />
            Register New Rover
          </Button>,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Enter the name of the rover",
            },
            "ref": null,
            "rendered": "Enter the name of the rover",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <Form
                as="form"
                id="nameForm"
                onSubmit={[Function]}
              >
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Rover name"
                  required={true}
                />
              </Form>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "form",
                "children": <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Rover name"
                  required={true}
                />,
                "id": "nameForm",
                "onSubmit": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "as": [Function],
                  "control": [Function],
                  "onChange": [Function],
                  "placeholder": "Rover name",
                  "required": true,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Button
                  as="button"
                  form="nameForm"
                  primary={true}
                  type="submit"
                >
                  Create
                </Button>,
                <Button
                  as="button"
                  onClick={[Function]}
                >
                  Cancel
                </Button>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "children": "Create",
                  "form": "nameForm",
                  "primary": true,
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Create",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "children": "Cancel",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Cancel",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "active": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          null,
          <Modal
            centered={true}
            closeOnDimmerClick={true}
            closeOnDocumentClick={false}
            dimmer={true}
            eventPool="Modal"
            onClose={[Function]}
            open={false}
            trigger={
              <Button
                as="button"
                onClick={[Function]}
                primary={true}
                style={
                  Object {
                    "marginLeft": "10px",
                  }
                }
              >
                <Icon
                  as="i"
                  name="plus"
                />
                Register New Rover
              </Button>
            }
          >
            <ModalHeader>
              Enter the name of the rover
            </ModalHeader>
            <ModalContent>
              <Form
                as="form"
                id="nameForm"
                onSubmit={[Function]}
              >
                <FormInput
                  as={[Function]}
                  control={[Function]}
                  onChange={[Function]}
                  placeholder="Rover name"
                  required={true}
                />
              </Form>
            </ModalContent>
            <ModalActions>
              <Button
                as="button"
                form="nameForm"
                primary={true}
                type="submit"
              >
                Create
              </Button>
              <Button
                as="button"
                onClick={[Function]}
              >
                Cancel
              </Button>
            </ModalActions>
          </Modal>,
          <Loader
            active={true}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "centered": true,
            "children": Array [
              <ModalHeader>
                Enter the name of the rover
              </ModalHeader>,
              <ModalContent>
                <Form
                  as="form"
                  id="nameForm"
                  onSubmit={[Function]}
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    onChange={[Function]}
                    placeholder="Rover name"
                    required={true}
                  />
                </Form>
              </ModalContent>,
              <ModalActions>
                <Button
                  as="button"
                  form="nameForm"
                  primary={true}
                  type="submit"
                >
                  Create
                </Button>
                <Button
                  as="button"
                  onClick={[Function]}
                >
                  Cancel
                </Button>
              </ModalActions>,
            ],
            "closeOnDimmerClick": true,
            "closeOnDocumentClick": false,
            "dimmer": true,
            "eventPool": "Modal",
            "onClose": [Function],
            "open": false,
            "trigger": <Button
              as="button"
              onClick={[Function]}
              primary={true}
              style={
                Object {
                  "marginLeft": "10px",
                }
              }
            >
              <Icon
                as="i"
                name="plus"
              />
              Register New Rover
            </Button>,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Enter the name of the rover",
              },
              "ref": null,
              "rendered": "Enter the name of the rover",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": <Form
                  as="form"
                  id="nameForm"
                  onSubmit={[Function]}
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    onChange={[Function]}
                    placeholder="Rover name"
                    required={true}
                  />
                </Form>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "form",
                  "children": <FormInput
                    as={[Function]}
                    control={[Function]}
                    onChange={[Function]}
                    placeholder="Rover name"
                    required={true}
                  />,
                  "id": "nameForm",
                  "onSubmit": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "as": [Function],
                    "control": [Function],
                    "onChange": [Function],
                    "placeholder": "Rover name",
                    "required": true,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Button
                    as="button"
                    form="nameForm"
                    primary={true}
                    type="submit"
                  >
                    Create
                  </Button>,
                  <Button
                    as="button"
                    onClick={[Function]}
                  >
                    Cancel
                  </Button>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "button",
                    "children": "Create",
                    "form": "nameForm",
                    "primary": true,
                    "type": "submit",
                  },
                  "ref": null,
                  "rendered": "Create",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "as": "button",
                    "children": "Cancel",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Cancel",
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "active": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
